java/src/main/java/com/google/appengine/tools/mapreduce/impl/shardedjob/ShardedJobRunner.java

```diff
@@ -84,28 +83,49 @@ public class ShardedJobRunner implements ShardedJobHandler {
   public static RetryerBuilder getRetryerBuilder() {
     return RetryerBuilder.newBuilder()
       .withWaitStrategy(WaitStrategies.exponentialWait(30_000, TimeUnit.MILLISECONDS))
+      .retryIfException(e -> {
+        if (e instanceof DatastoreException) {
+          return ((DatastoreException) e).isRetryable();
+        }
+        return false;
+      })
       .retryIfExceptionOfType(ApiProxyException.class)
-      .retryIfExceptionOfType(ConcurrentModificationException.class)
+      .retryIfExceptionOfType(ConcurrentModificationException.class) // don't think this is thrown by new datastore lib
       //.retryIfExceptionOfType(DatastoreFailureException.class)
       //.retryIfExceptionOfType(CommittedButStillApplyingException.class)
-     // .retryIfExceptionOfType(DatastoreTimeoutException.class)
+      // .retryIfExceptionOfType(DatastoreTimeoutException.class)
       .retryIfExceptionOfType(TransientFailureException.class)
       .retryIfExceptionOfType(TransactionalTaskException.class);
   }
 
   // NOTE: no StopStrategy set, must be set by the caller prior to build
   public static RetryerBuilder getRetryerBuilderAggressive() {
-      return RetryerBuilder.newBuilder()
-        .withWaitStrategy(WaitStrategies.exponentialWait(30_000, TimeUnit.MILLISECONDS))
-        .retryIfException(e ->
-          !(e instanceof RequestTooLargeException
-            || e instanceof ResponseTooLargeException
-            || e instanceof ArgumentException
-            || e instanceof DeadlineExceededException));
+    return RetryerBuilder.newBuilder()
+      .withWaitStrategy(WaitStrategies.exponentialWait(30_000, TimeUnit.MILLISECONDS))
+      .retryIfException(e ->
+        !(e instanceof RequestTooLargeException
+          || e instanceof ResponseTooLargeException
+          || e instanceof ArgumentException
+          || e instanceof DeadlineExceededException));
   }
 
 
 
+  public <T extends IncrementalTask> List<IncrementalTaskState<T>> lookupTasks(
+    final ShardedJobRunId jobId, final int taskCount, final boolean lenient) {
+    Transaction tx = datastore.newTransaction();
+    try {
+      List<IncrementalTaskState<T>> taskStates = new ArrayList<>();
```

Comment:
not a big deal, but if this is just read-only, why bother using a transaction at all? Can't we just use the `Datastore` class to read?

Reply:
in theory, you still want a consistent read; there are multiple datastore access happening underneath - getting all the task states, but then if they have large data, that may have been split and serialized into multiple blobs.